src/libcprover-cpp/api.cpp

```diff
@@ -137,6 +138,28 @@ std::unique_ptr<verification_resultt> api_sessiont::verify_model() const
 {
   PRECONDITION(implementation->model);
 
+  bool empty_result = preprocess_model();
+  if(empty_result)
+  {
+    return {};
+  }
+
+  return run_verifier();
+}
+
+void api_sessiont::read_goto_binary(std::string &file) const
+{
+  implementation->model = util_make_unique<goto_modelt>(
+    ::read_goto_binary(file, *implementation->message_handler).value());
+}
+
+bool api_sessiont::is_goto_binary(std::string &file) const
```

Comment:
:grey_question: Please consider whether having a separate `read_goto_binary` function and a `is_goto_binary` function is actually a good interface for a publicly facing API? This kind of implies that the file has to be opened and closed twice for the typical use case.

Reply:
We know that CBMC is not thread-safe in general, so I admit that this is a possibility, but bar from starting to add locks to various calls like that right now I don't know how to fix it.

I believe that adding locks/making it thread-safe probably falls outside of the scope of this PR. Do you have any suggestions on another way to achieve this, or did you want locks introduced?

Reply:
Thread safety is entirely unrelated issue here. Processes contain threads, the cbmc process has process-level (global) state which would be an issue if we wanted to use multiple threads within the cbmc process. This isn't what this issue is about, as cbmc does not use multiple threads and there is no plan to introduce them to my knowledge. However we should correctly support multi-processing (play-nicely) whether the processes involved are multiple instances of cbmc, or cbmc plus other unrelated user processes. I could go into hypothetical scenarios about how non-related processes could cause issues with the goto-binary in between the API calls but this isn't really the crux of the issue.

I think that having two separate API calls invites incorrect assumptions about how the two API calls relate to each other. Simply removing `is_goto_binary` from the public API would be one way to address this. Potential incorrect assumptions include -

- The assumption that if `is_goto_binary` returns true then the following call to `read_goto_binary` will succeed. This is incorrect even without considering concurrency / race condition scenarios. `is_goto_binary` only checks the header of the file. Therefore `read_goto_binary` could still fail due to error causing data after the header. Therefore the caller of `read_goto_binary` must handle any errors encountered during the API call even if a preceding call to `is_goto_binary` suceeded.
- The assumption that if we are going to call `read_goto_binary` then we must call `is_goto_binary` before hand. As I read the current implementation, `read_goto_binary` includes all the checks which `is_goto_binary` does. Therefore doing a precheck on `is_goto_binary` is unnecessary as it is redundant.

We could keep a publicly facing API call which does a check on whether a particular file is a goto-binary or not. But I don't see what the current advantage of doing so would be?

Reply:
Ah, I think I understand the problem better now. I will paraphrase the concern, to make sure I have understood it correctly - feel free to correct any points that are wrong:

> The issue with `is_goto_binary` and `read_goto_binary` is that the similar naming
> confuses because it implies an implicit sequencing between the two calls that sets
> up certain expectations that can be violated.

If my understanding is correct, would some documentation ameliorate those concerns?

With regards to the API call itself, I agree that it looks a bit out of place, but in general, my opinion is that the API as a whole should provide mechanism, not policy. I can easily picture the API call being part of some future scenarios, something like (pseudocode)

```c
if (is_goto_binary())
   extract_goto_binary_section(relro)
else if (is_elf_binary())
   extract_elf_binary_section(retro)
else
   // don't do anything for source files
```

I don't know how realistic this particular example might be, but scenarios that require some sort of querying of `goto-binary` properties fall within the purview of general API usage (if a bit underdeveloped right now), so I wouldn't want to remove from that capability.

Any thoughts?

Reply:
I think that we should shelve this discussion for potential continuation outside the forum of this PR review, as it was a point for consideration rather than a blocking comment.